kotlinx-coroutines-core/native/src/MultithreadedDispatchers.kt

```diff
@@ -73,38 +74,80 @@ private class MultiWorkerDispatcher(
     workersCount: Int
 ) : CloseableCoroutineDispatcher() {
     private val tasksQueue = Channel<Runnable>(Channel.UNLIMITED)
+    private val availableWorkers = Channel<CancellableContinuation<Runnable>>(Channel.UNLIMITED)
     private val workerPool = OnDemandAllocatingPool(workersCount) {
         Worker.start(name = "$name-$it").apply {
             executeAfter { workerRunLoop() }
         }
     }
 
+    /**
+     * (number of tasks - number of workers) * 2 + (1 if closed)
+     */
+    private val tasksAndWorkersCounter = atomic(0L)
+
+    private inline fun Long.isClosed() = this and 1L == 1L
+    private inline fun Long.hasTasks() = this >= 2
+    private inline fun Long.hasWorkers() = this < 0
+
     private fun workerRunLoop() = runBlocking {
-        // NB: we leverage tail-call optimization in this loop, do not replace it with
-        // .receive() without proper evaluation
-        for (task in tasksQueue) {
-            /**
-             * Any unhandled exception here will pass through worker's boundary and will be properly reported.
-             */
-            task.run()
+        while (true) {
+            val state = tasksAndWorkersCounter.getAndUpdate {
+                if (it.isClosed() && !it.hasTasks()) return@runBlocking
+                it - 2
+            }
+            if (state.hasTasks()) {
+                // we promised to process a task, and there are some
+                tasksQueue.receive().run()
+            } else {
+                try {
+                    suspendCancellableCoroutine {
+                        availableWorkers.trySend(it)
+                    }.run()
+                } catch (e: CancellationException) {
+                    /** we are cancelled from [close] and thus will never get back to this branch of code,
+                    but there may still be pending work, so we can't just exit here. */
```

Comment:
The code is written with this in mind: we can only end up on line 101 if we decremented the task count from a positive number, and it was positive because someone promised to provide a task (by incrementing the number). Notice that, if the counter was successfully incremented, there's no choice but to submit a task, even if the dispatcher is closed after that increment, so the task will always end up in the queue! As a result, the number of threads waiting on line 101 (now or in the near future) is always the same as the number of tasks that entered the channel (or will enter it shortly).

If this is true, all threads are guaranteed to leave this loop once no more tasks are left for them to execute. The stress test suggests that this does hold.

Does this make sense? Maybe there are some places where I could add a helpful comment?

Reply:
It seems like I'm missing something.

Consider the following scenario: 

* `close` is being in progress
* `close` cancels one of the waiting workers (W1), it catches `CancellationException` at 107, gets preempted
* W2 invokes `dispatch`: it increments the counter, gets preempted
* W1 gets scheduled, goes to the next loop iteration, observes the positive tasks counter, decrements it, goes to `receive`. There is no tasks in the channel (W2 got preempted), so it attempts to suspend and immediately throws `CancellationException` right from 101.
* W2 finished `dispatch` and sends task to the channel. Counter is already 0 and there are no worker who can execute it.

Reply:
> throws `CancellationException` right from 101 because it's already cancelled.

Why? Only the channel for personal communication is closed, the global one always lives, exactly because of the scenario you're describing.

Reply:
Oh, I see what I got mixed up. Thanks!